honeybadgermpc/reed_solomon.py

```diff
@@ -374,6 +375,69 @@ def get_results(self):
         return None, None
 
 
+class OptimalEncoder(Encoder):
+    def __init__(self, point):
+        self.encoder_van = VandermondeEncoder(point)
+        self.encoder_fft = FFTEncoder(point)
+        self.point = point
+        self.n = point.n
+        self.cores = psutil.cpu_count(logical=False)
+
+    def encode_one(self, data):
+        SetNumThreads(1)
+
+        # Parameters determined experimentally
+        if self.n < 8:
+            return self.encoder_van.encode_one(data)
+        else:
+            return self.encoder_fft.encode_one(data)
+
+    def encode_batch(self, data):
+        SetNumThreads(min(len(data), self.cores))
+
+        n = self.n
+        npow2 = n if n & (n - 1) == 0 else 2 ** n.bit_length()
+
+        # Check if n is close to the nearest power of 2
+        # In the worst case, n would be just one above a power of 2. For example, 65
+        # The nearest power of 2 greater than this is 128
+        # 128 - 65 = 63 > 128 / 4 = 32. This is bad. So we will use vandermonde here.
+        if npow2 - self.n > npow2 // 4 and self.n < 128 and len(data) > 256:
+            return self.encoder_van.encode_batch(data)
+        else:
+            return self.encoder_fft.encode_batch(data)
+
+
+class OptimalDecoder(Decoder):
+    def __init__(self, point):
+        self.decoder_van = VandermondeDecoder(point)
+        self.decoder_fft = FFTDecoder(point)
+        self.point = point
+        self.n = point.n
+        self.cores = psutil.cpu_count(logical=False)
+
+    def decode_one(self, z, data):
+        SetNumThreads(1)
+
+        # Parameters determined experimentally
+        if self.n < 8:
+            return self.decoder_van.decode_one(z, data)
+        else:
+            return self.decoder_fft.decode_one(z, data)
+
+    def decode_batch(self, z, data):
+        SetNumThreads(min(len(data), self.cores))
+
+        # Parameters determined experimentally
+        if self.n < 8:
```

Comment:
You should pull 8 out into a private static constant of the class. This will make it easier to test different parameters of this on different machines, as well as make the code easier to read.